## Introduction
Data science is a hot topic in today's world because it has helped companies and individuals extract value from data in unprecedented ways, particularly through ML.

In today's session, we're going to be going over some of the applications of data science and ML in predicting the housing market. More specifically, we'll be talking about supervised learning. 

### Our story for today:
Your cousin has made millions of dollars speculating in the housing market. He's offered to become business partners with you because of your interest in data science. He'll supply the money, and you'll supply models that predict how much various houses are worth.

You ask your cousin how he's predicted real estate values in the past, and he says it is just intuition. But more questioning reveals that he's identified price patterns from houses he has seen in the past, and he uses those patterns to make predictions for new houses he is considering.

At its core, machine learning works the same way. We'll give our model a labelled data set, and train it to find patterns in the data to make predictiosn for us. 

To begin this session, we'll start with a model called the Decision Tree. There are fancier models that give more accurate predictions. But decision trees are easy to understand, and they are the basic building block for some of the best models in data science.

**NB: this narrative^ was sourced from Kaggle's Intro to Machine Learning series**

## Decision Trees
A decision tree follows a set of if-else conditions to visualize the data and classify it according to the conditions. In the example shown below, the predicted price for any house under consideration is the historical average price of houses in the same category.

<img src = 'http://i.imgur.com/7tsb5b1.png'>

#### For nerds
To figure out which criteria are best to use, the decision tree algorithm seeks to minimize **impurity**. A popular impurity measure is the **Gini index**, calculated using the formula below:

<img src = "https://cdn-images-1.medium.com/max/300/0*pE3uG1i28u5ClQVQ.png">

where $p_i$ is the sum of squared probabilities of each class $i$. In this case $i = 2$ because we split our data into two classes: houses with more than 2 bedrooms, and houses with 2 or fewer bedrooms. 

Therefore, we can compute the 

### A better tree
We can capture more information and produce better predictions by introducing "splits" in our tree. With a split, we're able to use more features to make our pricing decision. This tree is "deeper" than our previous one because it has more splits.

We use the Gini score to determine which feature sits at the top of the tree and so on. 

<img src = "http://i.imgur.com/R3ywQsR.png">

We predict the price of any house by tracing through the decision tree, always picking the path corresponding to that house's characteristics. The predicted price for the house is at the bottom of the tree. The point at the bottom where we make a prediction is called a **leaf**.

## Pandas
Before we get to modelling though, we should learn how to examine our data. We'll use the Pandas library for this. Pandas is the primary tool data scientists/consultants use for exploring and manipulating data!

In [152]:
import pandas as pd

In [153]:
# Read the data 
# our data is already split into training and testing
# which we will discuss in a bit
train = pd.read_csv('train.csv', index_col='Id')

# data cleaning - dw abt this
# Remove rows with missing target information, 
# separate target from predictors
train.dropna(axis=0, subset=['SalePrice'], inplace=True)

# split into feature-space and target-space
y = train['SalePrice']
X = train.drop(['SalePrice'], axis=1)

In [154]:
from sklearn.model_selection import train_test_split

# we're going to split the data so we can test our model accuracy later on unseen data
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2, random_state = 42)

In [155]:
X_train.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1168.000000   951.000000    1168.000000  1168.000000  1168.000000   
mean     56.849315    70.343849   10689.642123     6.121575     5.584760   
std      42.531862    24.897021   10759.366198     1.367619     1.116062   
min      20.000000    21.000000    1300.000000     1.000000     1.000000   
25%      20.000000    59.000000    7587.250000     5.000000     5.000000   
50%      50.000000    70.000000    9600.000000     6.000000     5.000000   
75%      70.000000    80.000000   11700.000000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  ...  \
count  1168.000000   1168.000000  1162.000000  1168.000000  1168.000000  ...   
mean   1970.965753   1984.897260   103.771945   446.023973    45.152397  ...   
std      30.675495     20.733955   173.032238   459.070977   158.217499  ...   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000  ...   
25%    1953.000000   1966.000000     0.000000     0.000000     0.000000  ...   
50%    1972.000000   1994.000000     0.000000   384.500000     0.000000  ...   
75%    2001.000000   2004.000000   166.000000   721.000000     0.000000  ...   
max    2010.000000   2010.000000  1378.000000  5644.000000  1127.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  1168.000000  1168.000000  1168.000000    1168.000000  1168.000000   
mean    476.273973    95.946918    49.578767      21.839041     3.812500   
std     211.095373   129.685939    69.433580      62.083227    31.519664   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     341.000000     0.000000     0.000000       0.000000     0.000000   
50%     482.000000     0.000000    27.000000       0.000000     0.000000   
75%     576.000000   168.000000    74.000000       0.000000     0.000000   
max    1418.000000   857.000000   547.000000     552.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  1168.000000  1168.000000   1168.000000  1168.000000  1168.000000  
mean     15.407534     2.955479     51.267123     6.356164  2007.818493  
std      55.881148    41.648504    553.039684     2.670707     1.322639  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     5.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     480.000000   738.000000  15500.000000    12.000000  2010.000000  

[8 rows x 36 columns]

### Interpreting Description
**Describe will only show us a description of our numeric variables**
- count: shows number of rows with non-missing data
- mean: average
- std: standard deviation
- min, 25%, 50%, 75%, max: smallest value, the number at 25% is the number that is bigger than 25% of the values for that column of information, and so on so forth.

## Building your first ML model
There are a total of 79 features that we can use to estimate house prices. However, we will only be using a subset of these features.

In [157]:
len(X_train.columns) # number of features

79

In [156]:
X_train.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

### Choosing features
First, for simplicity, in this workshop, we'll only work with numeric data. Before we get to modelling, our goal is to find the "best" features in our data set to make our predictions with. In other words, we want to isolate the features with the most predictive power and eliminate the ones that are "noisy."

Methods for selecting features:
1. Intuition/domain expertise
2. Statistical methods
3. Machine learning

In [158]:
# get the numeric columns
numeric_cols = [col for col in X_train.columns if X_train[col].dtype == 'int64' or 
                X_train[col].dtype == 'float64']


# show number of numeric columns of data
print(f"We have {len(numeric_cols)} variables. Let's try to pare this down to a smaller subset!")

We have 36 variables. Let's try to pare this down to a smaller subset!


#### 1. Applying intuition/domain expertise
Utilising domain knowledge is probably the fastest way to identify variables, but it may also introduce bias. Perhaps what you know from the past is not representative of what shapes the housing market today! Nonetheless, it's a technique that can expedite analysis.

In [193]:
print(f'These are the columns we\'re choosing from:\n\n{numeric_cols}')

These are the columns we're choosing from:

['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold']


Suppose that we're consultants, and in our database & from past experience, we've identified the following variables as the most important
- LotArea: the lot area
- OverallQual: the quality rating of the property 
- GrLivArea: the gross living area of the property
- GarageArea: the area of the garage of the property
- MoSold: the month the property was sold 
- YrSold: the year the property was sold 
- TotRmsAbvGrd: the total number of rooms the property has (above ground)

We can now use these variables to build our decision tree model! In this process, we're not only going to select the relevant features outlined in the bulleted list, but we're also going to standardise the features. 

Standardisation is done by performing the following computation: for every vector $\overrightarrow x$ in our matrix of predictors $X$, compute $\frac{\overrightarrow x - \mu_x}{\sigma_x}$

Standardisation is important because it gives equal weights/importance to each feature so that no single feature steers model performance in one direction just because it is generally bigger than the other features.

In [160]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler

def preprocess1(X):
    """
    Preprocess our data by selecting the features
    outlined above and then standardising the feature-space
    
    Returns: new, standardised feature-space and our original
    target vector
    """
    X1 = X.copy()
    vars1 = ['LotArea','OverallQual','GrLivArea','GarageArea',
             'MoSold','YrSold','TotRmsAbvGrd']
    
    # select relevant features
    X1 = X1[vars1]
    
    # scale data
    scaler = StandardScaler()
    scaler.fit_transform(X1)
    
    return X1

# perform preprocessing
X_train1 = preprocess1(X_train)

# transform test set so we get Apples to Apples comparison
X_test1 = preprocess1(X_test) 

# define model: specify a number for random_state to ensure we get the same results each run
# (decision trees incorporate randomness)
dt_model = DecisionTreeRegressor(random_state=42)

# fit model
dt_model.fit(X_train1, y_train);

In [161]:
# a common error metric used in machine learning
from sklearn.metrics import mean_squared_error

# for math computations
import numpy as np 

# for fancy formatting w/ Latex
from IPython.display import display, Markdown


# make predictions
predictions1 = dt_model.predict(X_test1)
rmse1 = round(np.sqrt(mean_squared_error(predictions1,y_test)),2)

# print rmse
display(Markdown(rf"""$RMSE = \${rmse1}$"""))

$RMSE = \$41317.14$

#### Interpreting RMSE
RMSE represents the average difference between your models' predictions and the actual values. Our model was on average, incorrect by $\$41,317.14$. This is huge, so obviously our model is pretty bad, and we can't help our cousin, but we can apply a different strategy and see if that works!

#### 2. Statistical methods
Now we'll explore some statistical techniques for feature selection.

One popular statistical method is applying a variance threshold. This method looks at our feature-space $X$ and will remove variables with low (or zero) variance. The intuition behind this technique is that features with no/low variance will always provide the same information no matter what the actual outcome is, thus it provides us with no information.

Hence, we will remove such variables from our dataset.

In [162]:
from sklearn.feature_selection import VarianceThreshold
from sklearn.impute import SimpleImputer
from sklearn.ensemble import ExtraTreesRegressor

def preprocess2(X):
    """
    Preprocess our data by selecting the features
    outlined above and then standardising the feature-space
    
    Returns: new, standardised feature-space and our original
    target vector
    """
    X1 = X.copy()
    
    # reduce subset to numeric data
    X1 = X1[numeric_cols]
    
    # fill missing values
    imputer = SimpleImputer(strategy = 'median')
    X1 = imputer.fit_transform(X1)
    
    # scale data
    scaler = StandardScaler()
    scaler.fit_transform(X1)
    
    # select features that have > 0 variance
    selector = VarianceThreshold()
    selector.fit(X1)
    
    col_names = [numeric_cols[i] for i in range(len(numeric_cols)) if 
                 selector.get_support()[i]]
    
    X1 = selector.transform(X1)
    
    return X1, col_names

# perform preprocessing
X_train2, col_names = preprocess2(X_train)

# which columns were selected?
print(f"{len(col_names)} columns were selected:\n\n{col_names}")

# transform test set so we get Apples to Apples comparison
X_test2 = X_test[col_names]

# impute missing values
imputer = SimpleImputer(strategy = 'median')
X_test2 = imputer.fit_transform(X_test2)

# define model. specify a number for random_state to ensure same results each run
dt_model = DecisionTreeRegressor(random_state=42)

# fit model
dt_model.fit(X_train2, y_train);

36 columns were selected:

['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold']


In [163]:
# Make predictions
predictions2 = dt_model.predict(X_test2)
rmse2 = round(np.sqrt(mean_squared_error(predictions2,y_test)),2)

# print rmse
display(Markdown(rf"""$RMSE = \${rmse2}$"""))

$RMSE = \$38256.19$

Our predictions improved by roughly $\$3000$ on average! This tells us that using all of the data is better than our intuition. We can use this information to update our intuition and consulting database...but still, a terrible outcome.

#### 3. Applying Machine Learning

In [164]:
from sklearn.feature_selection import VarianceThreshold
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_selection import SelectFromModel

def preprocess3(X):
    """
    Preprocess our data by selecting the features
    outlined above and then standardising the feature-space
    
    Returns: new, standardised feature-space and our original
    target vector
    """
    X1 = X.copy()
    
    # reduce subset to numeric data
    X1 = X1[numeric_cols]
    
    # fill missing values
    imputer = SimpleImputer(strategy = 'median')
    X1 = imputer.fit_transform(X1)
    
    # scale data
    scaler = StandardScaler()
    scaler.fit_transform(X1)
    
    # select features
    # create & fit feature selection model
    rf = RandomForestRegressor(n_estimators = 500,random_state = 42,
                               max_depth = 10, n_jobs = -1,
                               criterion='mse')
    
    feature_select_mod = SelectFromModel(estimator = rf).fit(X1,y_train)
    
    # get names of good columns
    col_names = [numeric_cols[i] for i in range(len(numeric_cols)) if 
                 feature_select_mod.get_support()[i]]
    
    X2 = feature_select_mod.transform(X1)
    
    return X2, col_names

# perform preprocessing
X_train3, col_names = preprocess3(X_train)

# which columns were selected?
print(f"{len(col_names)} columns were selected:\n\n{col_names}")

# transform test set so we get Apples to Apples comparison
X_test3 = X_test[col_names]

# impute missing values
imputer = SimpleImputer(strategy = 'median')
X_test3 = imputer.fit_transform(X_test3)

# define model. specify a number for random_state to ensure same results each run
dt_model = DecisionTreeRegressor(random_state=42)

# fit model
dt_model.fit(X_train3, y_train);

6 columns were selected:

['OverallQual', 'BsmtFinSF1', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea']


In [165]:
# Make predictions
predictions3 = dt_model.predict(X_test3)
rmse3 = round(np.sqrt(mean_squared_error(predictions3,y_test)),2)

# print rmse
display(Markdown(rf"""$RMSE = \${rmse3}$"""))

$RMSE = \$37322.98$

Ok, wow, we did even better with only 6 variables! This demonstrates the power of feature selection, which can improve our ability to predict outcomes AND give us insight into variables that are the most important for predicting our outcomes. 

But why is the RMSE still so high? We can't get rich with our cousin if we don't reduce this value! We'll address this issue by considering models other than decision trees. The reason is: Decision trees are actually very bad IRL!!!

To quote ***The Elements of Statistical Learning***: 
>"Trees have one aspect that prevents them from being the ideal tool for predictive learning, namely **inaccuracy**." 

In other words, a decision tree is way too rigid to generalise!

## Testing other ML models
In this section, we're going to learn more about using the sci-kit learn library to create other models (like we did with our decision tree. When coding, this library is written as `sklearn`. Scikit-learn is easily the most popular library for modeling the types of data typically stored in our `pandas` DataFrames.

The steps to building and using a model are:
- Define: What type of model will it be? A tree? A linear model? Some other type of model? How will the parameters of the model be specified?
- Fit: Apply the model to our data to capture patterns
- Predict: Just what it sounds like
- Evaluate: Determine how accurate the model's predictions are on a general data set

Below, I'll show you a few examples of how to apply different machine learning models to our data. We're going to use `X_train2`, `X_train3`, and `X_test` to train our models and evaluate their predictive accuracy.

In [203]:
# import models from relevant libraries
from sklearn.metrics import mean_absolute_error
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import GradientBoostingRegressor

def test_models(X_train,X_test,y_train,y_test):
    """
    A function that will help us test a variety of specified models
    """

    # dictionary of models we want to test
    models = {'Neural Network':MLPRegressor(random_state=42, max_iter=500,
                                           hidden_layer_sizes(64)),
              'Linear Regression':LinearRegression(),
              'Elastic Net':ElasticNet(),
              'Ridge Regression':Ridge(),
              'Lasso Regression':Lasso(),
              'Support Vector Machine':SVR(),
              'Random Forest':RandomForestRegressor(n_estimators = 500,random_state = 42,
                                                    max_depth = 10, n_jobs = -1,
                                                    #convention to use 20 to smooth splits
                                                    criterion='mse',min_samples_split=20), 
              'Ada Boost':AdaBoostRegressor(n_estimators = 500,
                                            random_state = 42,loss = 'square'),
              'Gradient Boosting':GradientBoostingRegressor(n_estimators = 500,random_state = 42,
                                                            loss = 'huber',criterion = 'mse',
                                                            min_samples_split=20)}

    # print RMSE of each model
    for key in models.keys():
        
        # specify our model
        mod = models[key]
        
        # fit our model
        mod.fit(X_train, y_train)
        
        # generate predictions on unseen data
        predictions = mod.predict(X_test)
        
        # compute our error metric, the RMSE
        rmse = round(np.sqrt(mean_squared_error(predictions,y_test)),2)
        mae = round(mean_absolute_error(predictions,y_test),2)
        # display our error
        display(Markdown(rf"""{key} $RMSE = \${rmse}$"""))
        print('')

SyntaxError: positional argument follows keyword argument (2286929518.py, line 19)

You'll notice that some of the models I've specified have parameters explicitly written out. For instance, in the random forest model, you'll see that I've written `n_estimators = 500`. This specification is not necessary but can help in building more accurate models.  

Also, although models like `ElasticNet` do not have parameters specified in their parentheses, they can accept parameter specifications. It's just that they are not necessary. To find the parameters that each model can accept parameters, you can look into their respective docstrings.

In [199]:
import warnings
warnings.filterwarnings('ignore')

# run our function
# using ML selected features
test_models(X_train3,X_test3,y_train,y_test)

Neural Network $RMSE = \$48020.22$

Linear Regression $RMSE = \$39154.13$

Elastic Net $RMSE = \$39640.28$

Ridge Regression $RMSE = \$39152.29$

Lasso Regression $RMSE = \$39154.11$

Support Vector Machine $RMSE = \$88595.36$

Random Forest $RMSE = \$32342.56$

Ada Boost $RMSE = \$40931.74$

Gradient Boosting $RMSE = \$35074.78$

In [200]:
# using all numeric features
test_models(X_train2,X_test2,y_train,y_test)

Neural Network $RMSE = \$42017.48$

Linear Regression $RMSE = \$36846.58$

Elastic Net $RMSE = \$37284.35$

Ridge Regression $RMSE = \$36845.81$

Lasso Regression $RMSE = \$36846.3$

Support Vector Machine $RMSE = \$88652.4$

Random Forest $RMSE = \$31159.27$

Ada Boost $RMSE = \$38200.6$

Gradient Boosting $RMSE = \$28526.51$

In [201]:
# using features recommended by our consulting database/intuition
test_models(X_train1,X_test1,y_train,y_test)

Neural Network $RMSE = \$54207.74$

Linear Regression $RMSE = \$42186.04$

Elastic Net $RMSE = \$43306.24$

Ridge Regression $RMSE = \$42186.43$

Lasso Regression $RMSE = \$42186.05$

Support Vector Machine $RMSE = \$88651.94$

Random Forest $RMSE = \$36223.23$

Ada Boost $RMSE = \$37344.57$

Gradient Boosting $RMSE = \$34868.89$

What you'll notice is that our random forest model works the best in all cases! You'll also see that using all of the numeric features produces the best performance. However, even if that is the case, you may prefer to only use the subset of ML selected features because of the principle of parsimony -- i.e. simpler models are better. 